include/ginkgo/core/distributed/matrix.hpp

```diff
@@ -0,0 +1,232 @@
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_DISTRIBUTED_MATRIX_HPP_
+#define GKO_PUBLIC_CORE_DISTRIBUTED_MATRIX_HPP_
+
+
+#include <ginkgo/config.hpp>
+
+
+#if GINKGO_BUILD_MPI
+
+
+#include <ginkgo/core/base/dense_cache.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/base.hpp>
+#include <ginkgo/core/distributed/partition.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+namespace gko {
+namespace distributed {
+
+
+template <typename ValueType>
+class Vector;
+
+
+template <typename ValueType = double, typename LocalIndexType = int32,
+          typename GlobalIndexType = int64>
+class Matrix
+    : public EnableLinOp<Matrix<ValueType, LocalIndexType, GlobalIndexType>>,
+      public EnableCreateMethod<
+          Matrix<ValueType, LocalIndexType, GlobalIndexType>>,
+      public ConvertibleTo<
+          Matrix<next_precision<ValueType>, LocalIndexType, GlobalIndexType>>,
+      public DistributedBase {
+    friend class EnableCreateMethod<
+        Matrix<ValueType, LocalIndexType, GlobalIndexType>>;
+    friend class EnablePolymorphicObject<
+        Matrix<ValueType, LocalIndexType, GlobalIndexType>, LinOp>;
+    friend class Matrix<next_precision<ValueType>, LocalIndexType,
+                        GlobalIndexType>;
+
+public:
+    using value_type = ValueType;
+    using index_type = GlobalIndexType;
+    using local_index_type = LocalIndexType;
+    using global_index_type = GlobalIndexType;
+    using global_vector_type = gko::distributed::Vector<ValueType>;
+    using local_vector_type = typename global_vector_type::local_vector_type;
+    using local_matrix_type = gko::matrix::Csr<value_type, local_index_type>;
+
+    using EnableLinOp<Matrix>::convert_to;
+    using EnableLinOp<Matrix>::move_to;
+
+    void convert_to(Matrix<next_precision<value_type>, local_index_type,
+                           global_index_type>* result) const override;
+
+    void move_to(Matrix<next_precision<value_type>, local_index_type,
+                        global_index_type>* result) override;
+
+    /**
+     * Reads a square matrix from the device_matrix_data structure and a global
+     * partition.
+     *
+     * Both the number of rows and columns of the device_matrix_data is ignored.
+     * The global size of the final matrix is inferred only from the size row
+     * partition.
+     *
+     * @note The matrix data can contain entries for rows other than those owned
+     *        by the process. Entries for those rows are discarded.
+     *
+     * @param data  The device_matrix_data structure.
+     * @param partition  The global row and column partition.
```

Comment:
maybe we should get the term `symmetric` in there somehow?

Reply:
The matrix itself is not necessarily symmetric, so I don't think we should add it here. It would probably be more confusing.

Reply:
I was talking about the partition being symmetric, which is the term I recall seeing most of the time associated with this kind of partition.